candlepin/candlepin
# [m] Lock less binds phase 2

## Description
Today, all bind time pool operations are performed via a bind chain that passes around a shared bind context. The preProcess method performs most of the heavy weight operations, but does not perform any database operations untill the execute method is executed after the locks are placed.
In this PR, I have:
* created a DelayedPoolOp to capture the operations to perform once the locks are placed.
* split PostBindBonusPoolsOp into PostBindBonusPoolsOp and CheckBonusPoolsQuantitiesOp.
* Currently only PostBindBonusPoolsOp is lock-less in it's preProcessing. A following PR would tackle CheckBonusPoolsQuantitiesOp.
* PostBindBonusPoolsOp temporarily assigns the pools and entitlements in its preProcess step so pools and entitlements can be associated to create stack derived pools and keep the stack updated.


## Performance results:
### Test setup
* create an owner
* create a single pool with attributes virt-limit, stack_id and multi-entitlement.
* make sure quantity of pool > 1000
* create 100 hypervisor consumers
* using caracalla/bind.jmx , in 100 threads bind each consumer against the same pool. repeat this in a loop 10 times.
* The first bind should create stacked derived pools for each consumer, and following 9 binds should update the stack.

### Result
| Branch | Average  | Min | Max | Std Dev. | Error % |
| ------------- | ------------- |------------- |------------- |------------- |------------- |
| lock-less-2 | 5773  | 313 | 28667 | 4875.64 | 0.007 |
| lock-less-2 | 5924  | 384 | 33304 | 5032.88 | 0.008 |
| Master | 7196  | 282 | 32198 | 5255.77 | 0 |
| Master | 7941  | 318 | 34714 | 5656.93 | 0 |

### Errors observed in performance tests
w.r.t the 0.008% error observed in this branch, upon closer inspection of the logs, it was observed that this happens because the performance test was written in a way that we could have multiple requests from the same consumer trying to bind against the same pool. when that happens, we can get a ConstraintViolationException because the two requests have completed their pre-process step together, before a lock was placed on the consumer. This has been a known and acceptable risk of lock-less implementation, and the previously agreed upon solution was to fail all but one concurrent requests so the clients can try again later. Added an exception handling block in BindChain to do the same.
